### PR TITLE
feat(build): #0 add user input prompt

### PR DESCRIPTION
**Pull request by blaggacao**
- Enable to ask user for input

### CODE REVIEW

File: src/args/make-script/template.sh [hunk 1]
```diff
@@ -24,6 +24,24 @@ function prompt_user_for_confirmation {
...
+  local question="$2"
+
+  warn Please enter your input: \
+    info "$question" \
+    && if running_in_ci_cd_provider; then
+      critical Running on CI/CD provider, cannot provid user input
+    else
+      read -ep '>>> ' -r answer \
+        && eval $result="'$answer'" \
+        && if test "${answer}" != ''; then
```
Comment (kamadorueda) [inline src/args/make-script/template.sh:39]:
```suggestion
        && if test -n "${answer}"; then
```

File: src/args/make-script/template.sh [hunk 2]
```diff
@@ -24,6 +24,24 @@ function prompt_user_for_confirmation {
...
 
+function prompt_user_for_input {
+  local answer
+  local result="$1"
+  local question="$2"
+
+  warn Please enter your input: \
+    info "$question" \
+    && if running_in_ci_cd_provider; then
+      critical Running on CI/CD provider, cannot provid user input
```
Comment (kamadorueda) [inline src/args/make-script/template.sh:35]:
```suggestion
      critical Running on CI/CD provider, cannot provide user input
```

```suggestion
      critical Running on CI/CD provider, cannot ask for user input, there are no humans around
```

File: src/args/make-script/template.sh [hunk 3]
```diff
@@ -24,6 +24,24 @@ function prompt_user_for_confirmation {
...
+    else
+      read -ep '>>> ' -r answer \
+        && eval $result="'$answer'" \
```
Comment (kamadorueda) [inline src/args/make-script/template.sh:38]: it seems this `eval` is not needed because $result var is not used elsewhere

Comment (blaggacao) [reply src/args/make-script/template.sh:38]: ` local result="$1"`

the trick is: `prompt_user_for_input myresultvar "my question" && echo "$myresult"`

Maybe that can be resolved in a "safer" way while maintaining the ability to echo stuff to stdout?

Comment (kamadorueda) [reply src/args/make-script/template.sh:38]: I get the point, what about:


```bash
function prompt_user_for_input {
  local answer
  local default="${1}"

  warn Please type an answer to continue: \
    && if running_in_ci_cd_provider; then
      warn Running on CI/CD provider, assuming you would have answered: "${default}" \
        && echo "${default}"
    else
      read -ep '>>> ' -r answer \
        && echo "${answer}"
    fi
}
```

```bash
info Please enter your age
result=$(prompt_user_for_input "25")
info you are ${result} years old
```

the user is able to modify execution and the CI is able to work with the default value

Comment (blaggacao) [reply src/args/make-script/template.sh:38]: afaik, this would also capture `warn` outputs...

Comment (blaggacao) [reply src/args/make-script/template.sh]: The trade off is acceptable:

```
declare -i var
prompt_user_for_input var "question?"
```

... would have errored with a type mismatch telling something abount it's not an integer, if no int was supplied.

---

Now it only fails with a cryptic:

```
[INFO] Token mint quantity
++ prompt_user_for_input
++ local answer
++ local default=
++ warn Please enter your input:
++ echo_stderr '[WARNING]' Please enter your input:
++ echo '[WARNING]' Please enter your input:
[WARNING] Please enter your input:
++ running_in_ci_cd_provider
++ export CI
++ test -n ''
++ read -ep '>>> ' -r answer
>>> sht
++ test -z sht
++ echo sht
+ token_mint_qty=sht
/tmp/nix-shell.VaSzy8/tmplwxc2q_b/result-mintNativeToken-10prepareTransaction/bin/mint-native-tokens-prepare-transaction: line 120: sht: unbound variable
```

File: src/args/make-script/template.sh [hunk 4]
```diff
@@ -24,6 +24,22 @@ function prompt_user_for_confirmation {
...
+function prompt_user_for_input {
+  local answer
+  local default="${1:-}"
+
+  warn Please enter your input: \
+    && if running_in_ci_cd_provider; then
+      critical Running on CI/CD provider, cannot ask for user input, there are no humans around
+    else
+      read -ep '>>> ' -r answer \
+        && if test -z "${answer:default}"; then
```
Comment (kamadorueda) [inline src/args/make-script/template.sh:36]: this is not doing the fallback to $default

image

if you want that on CI/CD the script stops immediately, we can avoid locally having a default value and directly echo back the user input

```bash
function prompt_user_for_input {
  local answer

  warn Please enter your input: \
    && if running_in_ci_cd_provider; then
      critical Running on CI/CD provider, cannot ask for user input, there are no humans around
    else
      read -ep '>>> ' -r answer \
        && echo "${answer}"
    fi
}
```

or the correct syntax for what you try to do in this PR:

```bash
function prompt_user_for_input {
  local answer
  local default="${1:-}"

  warn Please enter your input: \
    && if running_in_ci_cd_provider; then
      critical Running on CI/CD provider, cannot ask for user input, there are no humans around
    else
      read -ep '>>> ' -r answer \
        && if test -z "${answer}"; then
          warn empty input, assuming default value of: "${default}" \
           && echo "${default}"
        else
           echo "${answer}"
        fi
    fi
}
```

File: src/args/make-script/template.sh (hunk 3 above)
Comment (kamadorueda) [reply src/args/make-script/template.sh]:
```
declare -i var
prompt_user_for_input var "question?"

Comment (kamadorueda) [reply src/args/make-script/template.sh]: this is a good interface IMHO

```
declare -i var
var=$(prompt_user_for_input "question?")
```

Comment (kamadorueda) [reply src/args/make-script/template.sh]: bash is not well suited for integer aritmetic jeje

I normally handle integers as strings, and operate them as integers with `$(( ... ))` as needed

File: src/args/make-script/template.sh (hunk 4 above)
Comment (blaggacao) [reply src/args/make-script/template.sh:36]: I almost got something working, but it looks the critical exit is not propagated:

Comment (blaggacao) [reply src/args/make-script/template.sh:38]: See changes: https://github.com/fluidattacks/makes/compare/cb81996c2ee006abc169f16b514bb918cc4f9555..b15c107d73ae12931032e9445eff3d6148e7ca77

**Comment by blaggacao** [comment]
This appears to be working now.

**Pull request merged**